src/op_utils.rs

```diff
@@ -6,33 +6,34 @@ use crate::reduction::EvalErr;
 
 pub fn check_arg_count<T: Allocator>(
     args: &Node<T>,
-    expected: i32,
+    expected: usize,
     name: &str,
 ) -> Result<(), EvalErr<T::Ptr>> {
-    let mut cnt = expected;
+    if arg_count(args, expected) != expected {
+        args.err(&format!(
+            "{} takes exactly {} argument{}",
+            name,
+            expected,
+            if expected == 1 { "" } else { "s" }
+        ))
+    } else {
+        Ok(())
+    }
+}
+
+pub fn arg_count<T: Allocator>(args: &Node<T>, return_early_if_exceeds: usize) -> usize {
+    let mut count = 0;
     // It would be nice to have a trait that wouldn't require us to copy every
     // node
     let mut ptr = args.clone();
-    loop {
-        match ptr.pair() {
-            Some((_, next)) => {
-                ptr = next.clone();
-            }
-            _ => {
-                return if cnt == 0 {
-                    Ok(())
-                } else {
-                    args.err(&format!(
-                        "{} takes exactly {} argument{}",
-                        name,
-                        expected,
-                        if expected == 1 { "" } else { "s" }
-                    ))
-                }
-            }
-        }
-        cnt -= 1;
+    while let Some((_, next)) = ptr.pair() {
```

Comment:
I'm unfamiliar with this. If `pair()` returns `None` so it doesn't match the `Some()`, does that make the whole `let Some((_, next)) = ...` expression itself be false? So is this essentially a short-hand for a match expression?

Reply:
I only know about it because `cargo clippy` suggested it! And it worked, so I left it. It used to be `loop` with `if let` (and `if let` is a shorthand match-expression).